runtime/near-vm-logic/src/logic.rs

```diff
@@ -812,6 +813,78 @@ impl<'a> VMLogic<'a> {
         self.internal_write_register(register_id, value_hash)
     }
 
+    /// Compute multiexponentiation for (sign, base, exp), serialized at bytes in LE notation.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// If `value_len-4 % 97 != 0`
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + alt_bn128_pairing_base + alt_bn128_pairing_byte * num_bytes`
+    pub fn alt_bn128_g1_multiexp(
+        &mut self,
+        value_len: u64,
+        value_ptr: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        fn weight(data: &[u8]) -> usize {
+            let mut log2 = 0;
+            let mut hamming = 0;
+            for i in 0..data.len() {
+                for j in 0..8 {
+                    if data[i] >> j == 1 {
+                        log2 = i * 8 + j;
+                        hamming += 1;
+                    }
+                }
+            }
+            log2 + hamming + 1
+        }
+
+        self.gas_counter.pay_base(alt_bn128_g1_multiexp_base)?;
+        let value_buf = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        // use manual gas metering here to avoid unpaid deserialization
+        let mut w_acc = 0;
+        for i in 0..(value_buf.len() - 4) / 97 {
+            w_acc += weight(&value_buf[4 + 1 + 64 + i * 97..4 + 1 + 64 + 32 + i * 97]);
+        }
+
+        self.gas_counter.pay_per_byte(alt_bn128_g1_multiexp_element, w_acc as u64)?;
+
+        let value = <Vec<(bool, G1, Fr)>>::try_from_slice(&value_buf)
+            .map_err(|e| HostError::GuestPanic { panic_msg: format!("{}", e) })?;
```

Comment:
`GuestPanic` is still not a correct error type, it is reserved for the cases when Wasm contract chooses to voluntarily panic. We need to define new error here, e.g. `HostError::AltBnU128DeserializationError` and `HostError::AltBnU128SerializationError`.